matchms/importing/load_from_msp.py

```diff
@@ -76,58 +76,72 @@ def parse_msp_file(filename: str) -> Generator[dict, None, None]:
 
             if contains_metadata(rline):
                 parse_metadata(rline, params)
-            else:
-                # Obtaining the masses and intensities
-                peak_pairs = get_peak_tuples(rline)
-
-                for peak in peak_pairs:
-                    mz, intensity = get_peak_values(peak)
-                    comment = get_peak_comment(peak)
-                    if comment is not None:
-                        peak_comments.update({mz: comment})
-
-                    peakscount += 1
-                    masses.append(mz)
-                    intensities.append(intensity)
-
-                # Obtaining the masses and intensities
-                if int(params['num peaks']) == peakscount:
-                    peakscount = 0
-                    yield {
-                        'params': (params),
-                        'm/z array': np.array(masses),
-                        'intensity array': np.array(intensities),
-                        'peak comments': peak_comments
-                    }
-
-                    params = {}
-                    masses = []
-                    intensities = []
-                    peak_comments = {}
-
-
-def get_peak_values(peak: str) -> Tuple[float, float]:
-    """ Get the m/z and intensity value from the line containing the peak information. """
-    splitted_line = peak.split(maxsplit=2)
-    mz = float(splitted_line[0].strip())
-    intensity = float(splitted_line[1].strip())
-    return mz, intensity
+                continue
+
+            mz, ints, comment = _parse_line_with_peaks(rline)
+
+            masses = np.append(masses, mz)
```

Comment:
Random kibitzing: `np.append` reallocates the array on every single call because `np.ndarray`'s buffer must exactly match its shape for all the fancy striding and indexing tricks to work efficiently. Most linear containers use an amortized pessimistic resizing algorithm, reserving more memory than they need to.

You want to use Python's `array` module to make `array.array('d')` which does use the amortized resizing rule but also supports the buffer protocol. You would grow the buffer in the `array.array`, and then at the very end pass it to `np.frombuffer` or `np.asarray` with dtype specified and `copy=False`. This will avoid re-copying those arrays and give you a NumPy array over that same memory.

Reply:
@mobiusklein This is very interesting, I didn't know that! Might be good to change it here since we kind of know the number of peaks to be read in advance given the num_peaks variable. So I think already allocating the right sized array would be the fastest?

Thank you very much for the hint!